src/util/bson/bson.hpp

```diff
@@ -32,41 +32,670 @@
+        return min_key_val;
+    }
+
+    explicit operator MaxKey() const
+    {
+        REALM_ASSERT(m_type == Bson::MAX_KEY);
+        return max_key_val;
+    }
+
+    explicit operator const IndexedMap<Bson>&&() const
+    {
+        REALM_ASSERT(m_type == Bson::DOCUMENT);
+        return std::move(*document_val);
+    }
+
+    explicit operator const std::vector<Bson>&&() const
+    {
+        REALM_ASSERT(m_type == Bson::ARRAY);
+        return std::move(*array_val);
+    }
+
+    size_t get_type() const noexcept
+    {
+        return m_type;
+    }
+
+    bool is_null() const;
+
+    bool operator==(const Bson& other) const
+    {
+        if (m_type != other.m_type) {
+            return false;
+        }
+
+        switch (m_type) {
+            case NONE:
+                return true;
+            case INT32:
+                return int32_val == other.int32_val;
+            case INT64:
+                return int64_val == other.int64_val;
+            case BOOL:
+                return bool_val == other.bool_val;
+            case DOUBLE:
+                return double_val == other.double_val;
+            case TIME:
+                return time_val == other.time_val;
+            case DATE:
+                return date_val == other.date_val;
+            case OID:
+                return oid_val == other.oid_val;
+            case DECIMAL:
+                return decimal_val == other.decimal_val;
+            case MAX_KEY:
+                return max_key_val == other.max_key_val;
+            case MIN_KEY:
+                return min_key_val == other.min_key_val;
+            case STRING:
+                return *string_val == *other.string_val;
+            case REGEX:
+                return *regex_val == *other.regex_val;
+            case BINARY:
+                return *binary_val == *other.binary_val;
+            case DOCUMENT:
+                return *document_val == *other.document_val;
+            case ARRAY:
+                return *array_val == *other.array_val;
+        }
+    }
+
+    bool operator!=(const Bson& other) const
+    {
+        return !(*this == other);
+    }
+
+private:
+    friend std::ostream& operator<<(std::ostream& out, const Bson& m);
+    template <typename T>
+    friend bool holds_alternative(const Bson& bson);
+
+    typedef enum Type {
+        NONE,
+        INT32,
+        INT64,
+        BOOL,
+        DOUBLE,
+        STRING,
+        BINARY,
+        TIME,
+        DATE,
+        OID,
+        DECIMAL,
+        REGEX,
+        MAX_KEY,
+        MIN_KEY,
+        DOCUMENT,
+        ARRAY
+    } Type;
+    Type m_type;
+
+    union {
+        int32_t int32_val;
+        int64_t int64_val;
+        bool bool_val;
+        double double_val;
+        time_t time_val;
+        Timestamp date_val;
+        ObjectId oid_val;
+        Decimal128 decimal_val;
+        MaxKey max_key_val;
+        MinKey min_key_val;
+        // ref types
+        RegularExpression* regex_val;
```

Comment:
These types (except maybe std::vector<Bson>?) don't need to be pointers. Instead they can be inline, and instead of `string_val = new std::string(...); delete string_val` you'd do `new (&string_val) std::string(...)` to construct and `string_val.~std::string()` to destroy.

Reply:
Ah, cool, this will clean things up. What is this syntax though?
`new (&string_val) std::string(...)`
I have not seen it before.

Reply:
[Placement new](https://en.cppreference.com/w/cpp/language/new#Placement_new)